design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
 
 ## Design Options
 
-### Option A
+### Adapter-Binder Approach
 
-Detail design option A here.
+Overview:
+
+* `Adapters` are bespoke per `Device` as well as per `Protocol`.
+* We term the `Hardware` as per the manufacturer specification to be `Device` + `Adapter`.
+* There is no separate binding.
+* There is no general concrete `Adapter` for any protocol. 
+* The process loop is kicked off by the `Environment`
+
+#### Adapters
+
+The problem with the current implementation is that the `Adapter` concept is too general and the `Binder` mechanism is too limited (1:1) to make linkup to a `Device` powerful or flexible enough.
+
+The `Adapters` in this design are per-protocol & per-device. So they can encapsulate all the logic necessary to drive the specific device that they wrap. 
+
+```python
+class StreamAdapterLinkum95(StreamAdapter):
```

Comment:
As I said yesterday, I think we can still design base classes for these adapters to mitigate one of the drawbacks you listed below. For example in the `Stream` case I suppose the server part with host and port, collecting stuff that has arrived and sending replies is always the same. The differences are then in how the concrete adapter handles the received input and how it assembles the reply.